core/flyout_vertical.js

```diff
@@ -0,0 +1,452 @@
+
+      this.addBlockListeners_(root, block, rect);
+
+      cursorY += blockHW.height + gaps[i];
+    } else if (item.type == 'button') {
+      this.initFlyoutButton_(item.button, cursorX, cursorY);
+      cursorY += item.button.height + gaps[i];
+    }
+  }
+};
+
+/**
+ * Determine if a drag delta is toward the workspace, based on the position
+ * and orientation of the flyout. This is used in determineDragIntention_ to
+ * determine if a new block should be created or if the flyout should scroll.
+ * @param {!goog.math.Coordinate} currentDragDeltaXY How far the pointer has
+ *     moved from the position at mouse down, in pixel units.
+ * @return {boolean} true if the drag is toward the workspace.
+ * @package
+ */
+Blockly.VerticalFlyout.prototype.isDragTowardWorkspace = function(
+    currentDragDeltaXY) {
+  var dx = currentDragDeltaXY.x;
+  var dy = currentDragDeltaXY.y;
+  // Direction goes from -180 to 180, with 0 toward the right and 90 on top.
+  var dragDirection = Math.atan2(dy, dx) / Math.PI * 180;
+
+  var range = this.dragAngleRange_;
+  if (this.toolboxPosition_ == Blockly.TOOLBOX_AT_LEFT) {
+    // Vertical at left.
+    if (dragDirection < range && dragDirection > -range) {
+      return true;
+    }
+  } else {
+    // Vertical at right.
+    if (dragDirection < -180 + range || dragDirection > 180 - range) {
+      return true;
+    }
+  }
+  return false;
+};
+
+/**
+ * Copy a block from the flyout to the workspace and position it correctly.
+ * @param {!Blockly.Block} originBlock The flyout block to copy.
+ * @return {!Blockly.Block} The new block in the main workspace.
+ * @private
+ */
+Blockly.VerticalFlyout.prototype.placeNewBlock_ = function(originBlock) {
+  var targetWorkspace = this.targetWorkspace_;
+  var svgRootOld = originBlock.getSvgRoot();
+  if (!svgRootOld) {
+    throw 'originBlock is not rendered.';
+  }
+  // Figure out where the original block is on the screen, relative to the upper
+  // left corner of the main workspace.
+  if (targetWorkspace.isMutator) {
+    var xyOld = this.workspace_.getSvgXY(/** @type {!Element} */ (svgRootOld));
+  } else {
+    var xyOld = Blockly.utils.getInjectionDivXY_(svgRootOld);
+  }
+
+  // Take into account that the flyout might have been scrolled horizontally
+  // (separately from the main workspace).
+  // Generally a no-op in vertical mode but likely to happen in horizontal
+  // mode.
+  var scrollX = this.workspace_.scrollX;
+  var scale = this.workspace_.scale;
+  xyOld.x += scrollX / scale - scrollX;
+
+  var targetMetrics = targetWorkspace.getMetrics();
+
+  // If the flyout is on the right side, (0, 0) in the flyout is offset to
+  // the right of (0, 0) in the main workspace.  Add an offset to take that
+  // into account.
+  var scrollX = 0;
+  if (this.toolboxPosition_ == Blockly.TOOLBOX_AT_RIGHT) {
+    scrollX = targetMetrics.viewWidth - this.width_;
+    // Scale the scroll (getSvgXY_ did not do this).
```

Comment:
In the old flyout code, there's a line here to reset scale to targetWorkspace.scale instead of this.workspace_.scale. Should that be here? If you started with scratch block's version, I think that would make sense.